### PR TITLE
HDPI-5486-Add-column-(region_id)-to-pcs_case-table

### PR DESCRIPTION
**Pull request by tvr-hmcts**
https://tools.hmcts.net/jira/browse/HDPI-5486

* Added a region_id to the pcs_case
* Changed the way the case management location is held and when it is held.
* Added the allocation of the court venue to the pcsCase

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/pcs/location/service/LocationReferenceService.java [hunk 1]
```diff
@@ -23,16 +24,17 @@ public class LocationReferenceService {
...
-    public List<CourtVenue> getCountyCourts(String authorisation, List<Integer> epimIds) {
+    public List<CourtVenue> getCourtVenues(List<Integer> epimIds) {
         if (Objects.isNull(epimIds) || epimIds.isEmpty()) {
             throw new IllegalArgumentException("epimIds cannot be null or empty");
         }
         String formattedEpimIds = formatEpimIds(epimIds);
         log.debug("Getting County courts from /refdata/location/court-venues for EpimIds {}", formattedEpimIds);
-        return locationReferenceApi.getCountyCourts(authorisation, authTokenGenerator.generate(),
-                formattedEpimIds, countyCourtTypeId);
+        return locationReferenceApi.getCourtVenues(authTokenGenerator.generate(),
```
Comment (sadmanrahh) [inline src/main/java/uk/gov/hmcts/reform/pcs/location/service/LocationReferenceService.java]: The token order looks swapped here. LocationReferenceApi expects the IDAM token as Authorization and the S2S token as ServiceAuthorization, but this passes them the other way around. Can we swap these and update the test to verify the correct order?

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/page/makeaclaim/StatementOfTruth.java [hunk 2]
```diff
@@ -42,4 +51,13 @@ public void addTo(PageBuilder pageBuilder) {
             .done()
             .label("statementOfTruth-saveAndReturn", CommonPageContent.SAVE_AND_RETURN);
     }
+
+    private AboutToStartOrSubmitResponse<PCSCase, State> midEvent(CaseDetails<PCSCase, State> details,
+                                                                  CaseDetails<PCSCase, State> detailsBefore) {
+        if (allocateRegionId) {
+            pcsCaseService.allocateRegionId(details.getData());
+        }
+        return AboutToStartOrSubmitResponse.<PCSCase, State>builder().build();
```
Comment (sadmanrahh) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/page/makeaclaim/StatementOfTruth.java]: This mid-event mutates the case data but returns no data in the response. Can we return .data(details.getData()) so the allocated regionId/caseManagementLocationNumber is written back to CCD before submission?

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/PcsCaseService.java [hunk 3]
```diff
@@ -90,13 +93,20 @@ public PcsCaseEntity loadCase(long caseReference) {
...
+        pcsCase.setCaseManagementLocationNumber(caseManagementLocation);
+    }
+
+    public void allocateRegionId(PCSCase pcsCase) {
+        allocateCaseManagementLocation(pcsCase);
+        if (pcsCase.getCaseManagementLocationNumber() != null) {
+            List<CourtVenue> courtVenues = locationReferenceService
+                .getCourtVenues(List.of(pcsCase.getCaseManagementLocationNumber()));
+            if (courtVenues != null && !courtVenues.isEmpty()) {
+                pcsCase.setRegionId(courtVenues.getFirst().epimmsId());
```
Comment (sadmanrahh) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/PcsCaseService.java]: This appears to save the court EPIMS ID as regionId. Should this use the actual region_id from location-reference instead, since CaseLocation.region expects the region value rather than the base location?

File: src/main/java/uk/gov/hmcts/reform/pcs/postcodecourt/service/PostCodeCourtService.java [hunk 4]
```diff
@@ -29,18 +29,16 @@ public PostCodeCourtService(PostCodeCourtRepository postCodeCourtRepository,
...
         List<PostCodeCourtEntity> results = getPostcodeCourtMappings(postCode);
-
         if (results.isEmpty()) {
-            log.error("EpimId not found, Case management location couldn't be allocated for postcode: {}", postCode);
+            log.error("EpimId not found, Court management location couldn't be allocated for postcode: {}", postCode);
             return null;
         }
         if (results.size() > 1) {
-            log.error("Multiple EpimId's found, Case management location not allocated for postcode: {}", postCode);
+            log.error("Multiple EpimId's found, Court management location not allocated for postcode: {}", postCode);
```
Comment (sadmanrahh) [inline src/main/java/uk/gov/hmcts/reform/pcs/postcodecourt/service/PostCodeCourtService.java]: Minor - maybe should be 'EpimIds' instead

File: src/test/java/uk/gov/hmcts/reform/pcs/ccd/service/PcsCaseServiceTest.java [hunk 5]
```diff
@@ -93,7 +97,8 @@ void setUp() {
             addressMapper,
             caseLinkService,
             caseFlagService,
-            postCodeCourtService
+            postCodeCourtService,
+            locationReferenceService
         );
     }
```
Comment (sadmanrahh) [inline src/test/java/uk/gov/hmcts/reform/pcs/ccd/service/PcsCaseServiceTest.java:104]: Can we add a test for when location-reference returns an empty list? That would confirm regionId remains null and the submit flow does not fail unexpectedly.

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/PcsCaseService.java [hunk 6]
```diff
@@ -60,7 +62,8 @@ public void createMainClaimOnCase(long caseReference, PCSCase pcsCase) {
         pcsCaseEntity.addDocuments(documentEntities);
         claimEntity.addClaimDocuments(documentEntities);
         pcsCaseEntity.addClaim(claimEntity);
-
+        pcsCaseEntity.setCaseManagementLocation(pcsCase.getCaseManagementLocationNumber());
```
Comment (benouaer) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/PcsCaseService.java:65]: is it possible to rename this as epimsId or baseLocation, as caseManagement location is an object that includes both region and baseLocation

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/PcsCaseService.java (hunk 3 above)
Comment (tvr-hmcts) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/PcsCaseService.java]: Ah thank you.  Caught up and missed the obvious.

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/PcsCaseService.java (hunk 6 above)
Comment (tvr-hmcts) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/PcsCaseService.java:65]: I found this quite confusing also TBH  I can rename.  BaseLocation?   There appears to be bot epims_id and epimms_id used within the api - which is also confusing.

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/page/makeaclaim/StatementOfTruth.java [hunk 7]
```diff
@@ -42,4 +51,15 @@ public void addTo(PageBuilder pageBuilder) {
             .done()
             .label("statementOfTruth-saveAndReturn", CommonPageContent.SAVE_AND_RETURN);
     }
+
+    private AboutToStartOrSubmitResponse<PCSCase, State> midEvent(CaseDetails<PCSCase, State> details,
+                                                                  CaseDetails<PCSCase, State> detailsBefore) {
+        if (allocateRegionId) {
+            pcsCaseService.allocateRegionId(details.getData());
```
Comment (benouaer) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/page/makeaclaim/StatementOfTruth.java:58]: can we set the caseManagementLocation here, i.e. both region and baseLocation, because I think CCD will check that rather than regionId to grant access to the case

**Comment by tvr-hmcts** [comment]
There is a problem with the pipeline workspace.

Closing this so to open a new PR (easier than fixing the workspace it appears as I do not have permission to do so) ... then will reference back to here from the new one.

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/page/makeaclaim/StatementOfTruth.java (hunk 7 above)
Comment (tvr-hmcts) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/page/makeaclaim/StatementOfTruth.java:58]: Ah it does.  allocatedRegionId calls that internally and sets it on the PCSCase also.
